jolt-core/src/jolt/vm/read_write_memory.rs

```diff
@@ -355,354 +383,549 @@ impl<F: JoltField, C: CommitmentScheme<Field = F>> ReadWriteMemory<F, C> {
+                        || sw_flag[i].is_one()
                     {
-                        read_tuples.insert((a, v, t_final[a as usize]));
-                        write_tuples.insert((a, v, timestamp));
-                    }
-
-                    v_read[RS1].push(v);
-                    t_read[RS1].push(t_final[a as usize]);
-                    t_final[a as usize] = timestamp;
-                }
-                MemoryOp::Write(a, v) => {
-                    panic!("Unexpected rs1 MemoryOp::Write({}, {})", a, v);
-                }
-            };
-
-            match step[RS2] {
-                MemoryOp::Read(a, v) => {
-                    assert!(a < REGISTER_COUNT);
-                    debug_assert_eq!(v, v_final[a as usize]);
-
-                    #[cfg(test)]
-                    {
-                        read_tuples.insert((a, v, t_final[a as usize]));
-                        write_tuples.insert((a, v, timestamp));
-                    }
-
-                    v_read[RS2].push(v);
-                    t_read[RS2].push(t_final[a as usize]);
-                    t_final[a as usize] = timestamp;
-                }
-                MemoryOp::Write(a, v) => panic!("Unexpected rs2 MemoryOp::Write({}, {})", a, v),
-            };
-
-            match step[RD] {
-                MemoryOp::Read(a, v) => panic!("Unexpected rd MemoryOp::Read({}, {})", a, v),
-                MemoryOp::Write(a, v_new) => {
-                    assert!(a < REGISTER_COUNT);
-                    let v_old = v_final[a as usize];
-
-                    #[cfg(test)]
-                    {
-                        read_tuples.insert((a, v_old, t_final[a as usize]));
-                        write_tuples.insert((a, v_new, timestamp + 1));
+                        match step[RAM_1_INDEX] {
+                            MemoryOp::Read(a, v) => {
+                                assert!(a >= program_io.memory_layout.input_start);
+                                let remapped_a = remap_address(a, &program_io.memory_layout);
+                                let remapped_a_index = remap_address_index(remapped_a);
+                                debug_assert_eq!(v, v_final_ram[remapped_a_index]);
+
+                                #[cfg(test)]
+                                {
+                                    r_tuples_ram.lock().unwrap().insert((
+                                        remapped_a,
+                                        v,
+                                        t_final_ram[remapped_a_index],
+                                    ));
+                                }
+                                #[cfg(test)]
+                                {
+                                    w_tuples_ram
+                                        .lock()
+                                        .unwrap()
+                                        .insert((remapped_a, v, timestamp));
+                                }
```

Comment:
nit (here and elsewhere)
```suggestion
                                #[cfg(test)]
                                {
                                    r_tuples_ram.lock().unwrap().insert((
                                        remapped_a,
                                        v,
                                        t_final_ram[remapped_a_index],
                                    ));
                                    w_tuples_ram
                                        .lock()
                                        .unwrap()
                                        .insert((remapped_a, v, timestamp));
                                }
```

Reply:
I separated it into two blocks to release mutex locks immediately after each write, but I think it should be okay to keep them in a single block for readability (since it's just tests).